Properly escape the names of Hive columns

bradenmacdonald opened the pull request:
**Description**: In [newer versions of Hive](http://mail-archives.apache.org/mod_mbox/hive-commits/201503.mbox/%3C20150310220045.68294AC02E7@hades.apache.org%3E), `date` and `interval` are keywords. This change escapes column names and partition key names when generating Hive DML queries.

Without this fix, the following error can occur when running the pipeline:
```
Failed to recognize predicate 'DATE'. Failed rule: 'identifier' in table name
```

This fix should be fully compatible with any version of Hive.

**JIRA**: [OSPR-718](https://openedx.atlassian.net/browse/OSPR-718)

**Merge deadline**: Not an edX partner, so best effort.

**Reviewers**: @mtyaka and TBD

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! It looks like you're a member of a company that does contract work for edX. If you're doing this work as part of a paid contract with edX, you should talk to edX about who will review this pull request. If this work is not part of a paid contract with edX, then you should ensure that there is an OSPR issue to track this work in [JIRA](https://openedx.atlassian.net), so that we don't lose track of your pull request. 

To automatically create an OSPR issue for this pull request, just visit this link: http://openedx-webhooks.herokuapp.com/github/process_pr?repo=edx%2Fedx-analytics-pipeline&number=127

mtyaka commented:
:+1:

edx-webhook commented:
Thanks for the pull request, @bradenmacdonald! I've created [OSPR-718](https://openedx.atlassian.net/browse/OSPR-718) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams. 

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ('this must be merged by XX date', and why that is)
- partner information ('this is a course on edx.org')
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the Github pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

We can't start reviewing your pull request until you've added yourself to the [AUTHORS](https://github.com/edx/edx-analytics-pipeline/blob/master/AUTHORS) file. Please see the [CONTRIBUTING](https://github.com/edx/edx-platform/blob/master/CONTRIBUTING.rst) file for more information.

sarina commented:
@bradenmacdonald can you add yourself to AUTHORS for this repo?

bradenmacdonald commented:
@sarina  Done :)

mulby commented:
There are still a few date column references in `user_activity.py` that need to be updated see the queries in CourseActivityDailyTask and CourseActivityMonthlyTask.

mulby commented:
also the stuff near line 83 of database_imports.py

mulby commented:
Also line 203 of location_per_course.py

mulby commented:
Took a pass, generally looks good. I found a few other Hive queries that contain "date" without backticks and enumerated them in comments above.

bradenmacdonald commented:
@mulby Ok, I rebased and added in those extra "date" references you noticed. I wasn't sure what you meant in database_imports.py, but I saw a generic unescaped column name input there so I escaped it.

mulby commented:
Thanks @bradenmacdonald, this passed acceptance tests, so I'm merging it in to master.

prabhanshu commented:
@mulby and @bradenmacdonald  

We are facing this issue while setting up pipeline. We are using AWS EMR

2016-10-11 12:04:30,336 INFO 13654 [luigi-interface] hadoop.py:219 - FAILED: ParseException line 3:4 Failed to recognize predicate 'date'. Failed rule: 'identifier' in column specification
2016-10-11 12:04:31,194 ERROR 13654 [luigi-interface] worker.py:304 - [pid 13654] Worker Worker(salt=290289791, host=ip-xx-xx-xx-xx, username=hadoop, pid=13654) failed    HiveTableFromParameterQueryTask(warehouse_path=s3://edx-analytics-logs-test/warehouse/hive/, insert_query=
            SELECT
                ce.`date`,
                ce.course_id,
                IF(p.gender != '', p.gender, NULL),
                SUM(ce.at_end),
                COUNT(ce.user_id)
            FROM course_enrollment ce
            LEFT OUTER JOIN auth_userprofile p ON p.user_id = ce.user_id
            GROUP BY
                ce.`date`,
                ce.course_id,
                IF(p.gender != '', p.gender, NULL)
        , table=course_enrollment_gender_daily, columns=(('date', 'STRING'), ('course_id', 'STRING'), ('gender', 'STRING'), ('count', 'INT'), ('cumulative_count', 'INT')), partition=dt=2016-10-11)
Traceback (most recent call last):
  File "/var/lib/analytics-tasks/7e20d842-7d8e-4f86-b426-8c771d8bde4c/venv/local/lib/python2.7/site-packages/luigi/worker.py", line 292, in _run_task
    task.run()
  File "/var/lib/analytics-tasks/7e20d842-7d8e-4f86-b426-8c771d8bde4c/venv/local/lib/python2.7/site-packages/luigi/hadoop.py", line 552, in run
    self.job_runner().run_job(self)
  File "/var/lib/analytics-tasks/7e20d842-7d8e-4f86-b426-8c771d8bde4c/venv/local/lib/python2.7/site-packages/edx/analytics/tasks/util/hive.py", line 353, in run_job
    return super(OverwriteAwareHiveQueryRunner, self).run_job(job)
  File "/var/lib/analytics-tasks/7e20d842-7d8e-4f86-b426-8c771d8bde4c/venv/local/lib/python2.7/site-packages/luigi/hive.py", line 359, in run_job
    return luigi.hadoop.run_and_track_hadoop_job(arglist)
  File "/var/lib/analytics-tasks/7e20d842-7d8e-4f86-b426-8c771d8bde4c/venv/local/lib/python2.7/site-packages/luigi/hadoop.py", line 264, in run_and_track_hadoop_job
    return track_process(arglist, tracking_url_callback, env)
  File "/var/lib/analytics-tasks/7e20d842-7d8e-4f86-b426-8c771d8bde4c/venv/local/lib/python2.7/site-packages/luigi/hadoop.py", line 248, in track_process
    raise HadoopJobError(message + 'Also, no tracking url found.', out, err)

bradenmacdonald commented:
@prabhanshu Sounds the the version of EMR you are using and the version of `edx-analytics-pipeline` are not compatible, but I'm not sure. GitHub isn't really a support forum, so please ask on https://groups.google.com/forum/#!forum/openedx-analytics if you're still stuck.
